ngx_http_ssl: add ssl_trusted_ca_path config option.
Directory analogue of ssl_trusted_certificate to verify mTLS clients.
Certs will not be sent to the client.
--- nginx.orig/src/event/ngx_event_openssl.c
+++ nginx/src/event/ngx_event_openssl.c
@@ -949,27 +949,28 @@ ngx_ssl_client_certificate(ngx_conf_t *c
 
 ngx_int_t
 ngx_ssl_trusted_certificate(ngx_conf_t *cf, ngx_ssl_t *ssl, ngx_str_t *cert,
-    ngx_int_t depth)
+    ngx_str_t *ca_path, ngx_int_t depth)
 {
     SSL_CTX_set_verify(ssl->ctx, SSL_CTX_get_verify_mode(ssl->ctx),
                        ngx_ssl_verify_callback);
 
     SSL_CTX_set_verify_depth(ssl->ctx, depth);
 
-    if (cert->len == 0) {
+    if (cert->len == 0 && ca_path->len == 0) {
         return NGX_OK;
     }
 
-    if (ngx_conf_full_name(cf->cycle, cert, 1) != NGX_OK) {
+    if (cert->len && ngx_conf_full_name(cf->cycle, cert, 1) != NGX_OK) {
         return NGX_ERROR;
     }
 
-    if (SSL_CTX_load_verify_locations(ssl->ctx, (char *) cert->data, NULL)
+    if (SSL_CTX_load_verify_locations(ssl->ctx, cert->len ? (char *) cert->data: NULL, ca_path->len ? (char *) ca_path->data : NULL)
         == 0)
     {
         ngx_ssl_error(NGX_LOG_EMERG, ssl->log, 0,
-                      "SSL_CTX_load_verify_locations(\"%s\") failed",
-                      cert->data);
+                      "trusted_certificate SSL_CTX_load_verify_locations(\"%s\", \"%s\") failed",
+                      cert->data,
+                      ca_path->data);
         return NGX_ERROR;
     }
 
--- nginx.orig/src/event/ngx_event_openssl.h
+++ nginx/src/event/ngx_event_openssl.h
@@ -203,7 +203,7 @@ ngx_int_t ngx_ssl_ciphers(ngx_conf_t *cf
 ngx_int_t ngx_ssl_client_certificate(ngx_conf_t *cf, ngx_ssl_t *ssl,
     ngx_str_t *cert, ngx_int_t depth);
 ngx_int_t ngx_ssl_trusted_certificate(ngx_conf_t *cf, ngx_ssl_t *ssl,
-    ngx_str_t *cert, ngx_int_t depth);
+    ngx_str_t *cert, ngx_str_t *ca_path, ngx_int_t depth);
 ngx_int_t ngx_ssl_crl(ngx_conf_t *cf, ngx_ssl_t *ssl, ngx_str_t *crl);
 ngx_int_t ngx_ssl_stapling(ngx_conf_t *cf, ngx_ssl_t *ssl,
     ngx_str_t *file, ngx_str_t *responder, ngx_uint_t verify);
--- nginx.orig/src/http/modules/ngx_http_grpc_module.c
+++ nginx/src/http/modules/ngx_http_grpc_module.c
@@ -36,6 +36,7 @@ typedef struct {
     ngx_str_t                  ssl_ciphers;
     ngx_uint_t                 ssl_verify_depth;
     ngx_str_t                  ssl_trusted_certificate;
+    ngx_str_t                  ssl_trusted_ca_path;
     ngx_str_t                  ssl_crl;
     ngx_array_t               *ssl_conf_commands;
 #endif
@@ -416,6 +417,13 @@ static ngx_command_t  ngx_http_grpc_comm
       offsetof(ngx_http_grpc_loc_conf_t, ssl_trusted_certificate),
       NULL },
 
+    { ngx_string("grpc_ssl_trusted_ca_path"),
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_TAKE1,
+      ngx_conf_set_str_slot,
+      NGX_HTTP_LOC_CONF_OFFSET,
+      offsetof(ngx_http_grpc_loc_conf_t, ssl_trusted_ca_path),
+      NULL },
+
     { ngx_string("grpc_ssl_crl"),
       NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_TAKE1,
       ngx_conf_set_str_slot,
@@ -4360,6 +4368,7 @@ ngx_http_grpc_create_loc_conf(ngx_conf_t
      *     conf->ssl_protocols = 0;
      *     conf->ssl_ciphers = { 0, NULL };
      *     conf->ssl_trusted_certificate = { 0, NULL };
+     *     conf->ssl_trusted_ca_path = { 0, NULL };
      *     conf->ssl_crl = { 0, NULL };
      */
 
@@ -4493,6 +4502,8 @@ ngx_http_grpc_merge_loc_conf(ngx_conf_t
                               prev->ssl_verify_depth, 1);
     ngx_conf_merge_str_value(conf->ssl_trusted_certificate,
                               prev->ssl_trusted_certificate, "");
+    ngx_conf_merge_str_value(conf->ssl_trusted_ca_path,
+                              prev->ssl_trusted_ca_path, "");
     ngx_conf_merge_str_value(conf->ssl_crl, prev->ssl_crl, "");
 
     ngx_conf_merge_ptr_value(conf->upstream.ssl_certificate,
@@ -4995,14 +5006,15 @@ ngx_http_grpc_set_ssl(ngx_conf_t *cf, ng
     }
 
     if (glcf->upstream.ssl_verify) {
-        if (glcf->ssl_trusted_certificate.len == 0) {
+        if (glcf->ssl_trusted_certificate.len == 0 || glcf->ssl_trusted_ca_path.len == 0) {
             ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
-                      "no grpc_ssl_trusted_certificate for grpc_ssl_verify");
+                      "no grpc_ssl_trusted_certificate or grpc_ssl_trusted_ca_path for grpc_ssl_verify");
             return NGX_ERROR;
         }
 
         if (ngx_ssl_trusted_certificate(cf, glcf->upstream.ssl,
                                         &glcf->ssl_trusted_certificate,
+										&glcf->ssl_trusted_ca_path,
                                         glcf->ssl_verify_depth)
             != NGX_OK)
         {
--- nginx.orig/src/http/modules/ngx_http_proxy_module.c
+++ nginx/src/http/modules/ngx_http_proxy_module.c
@@ -123,6 +123,7 @@ typedef struct {
     ngx_str_t                      ssl_ciphers;
     ngx_uint_t                     ssl_verify_depth;
     ngx_str_t                      ssl_trusted_certificate;
+    ngx_str_t                      ssl_trusted_ca_path;
     ngx_str_t                      ssl_crl;
     ngx_array_t                   *ssl_conf_commands;
 #endif
@@ -743,6 +744,13 @@ static ngx_command_t  ngx_http_proxy_com
       offsetof(ngx_http_proxy_loc_conf_t, ssl_trusted_certificate),
       NULL },
 
+    { ngx_string("proxy_ssl_trusted_ca_path"),
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_TAKE1,
+      ngx_conf_set_str_slot,
+      NGX_HTTP_LOC_CONF_OFFSET,
+      offsetof(ngx_http_proxy_loc_conf_t, ssl_trusted_ca_path),
+      NULL },
+
     { ngx_string("proxy_ssl_crl"),
       NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_TAKE1,
       ngx_conf_set_str_slot,
@@ -3350,6 +3358,7 @@ ngx_http_proxy_create_loc_conf(ngx_conf_
      *     conf->ssl_protocols = 0;
      *     conf->ssl_ciphers = { 0, NULL };
      *     conf->ssl_trusted_certificate = { 0, NULL };
+     *     conf->ssl_trusted_ca_path = { 0, NULL };
      *     conf->ssl_crl = { 0, NULL };
      */
 
@@ -3751,6 +3760,8 @@ ngx_http_proxy_merge_loc_conf(ngx_conf_t
                               prev->ssl_verify_depth, 1);
     ngx_conf_merge_str_value(conf->ssl_trusted_certificate,
                               prev->ssl_trusted_certificate, "");
+    ngx_conf_merge_str_value(conf->ssl_trusted_ca_path,
+                              prev->ssl_trusted_ca_path, "");
     ngx_conf_merge_str_value(conf->ssl_crl, prev->ssl_crl, "");
 
     ngx_conf_merge_ptr_value(conf->upstream.ssl_certificate,
@@ -5041,14 +5052,15 @@ ngx_http_proxy_set_ssl(ngx_conf_t *cf, n
     }
 
     if (plcf->upstream.ssl_verify) {
-        if (plcf->ssl_trusted_certificate.len == 0) {
+        if (plcf->ssl_trusted_certificate.len == 0 || plcf->ssl_trusted_ca_path.len == 0) {
             ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
-                      "no proxy_ssl_trusted_certificate for proxy_ssl_verify");
+                      "no proxy_ssl_trusted_certificate or proxy_ssl_trusted_ca_path for proxy_ssl_verify");
             return NGX_ERROR;
         }
 
         if (ngx_ssl_trusted_certificate(cf, plcf->upstream.ssl,
                                         &plcf->ssl_trusted_certificate,
+                                        &plcf->ssl_trusted_ca_path,
                                         plcf->ssl_verify_depth)
             != NGX_OK)
         {
--- nginx.orig/src/http/modules/ngx_http_ssl_module.c
+++ nginx/src/http/modules/ngx_http_ssl_module.c
@@ -178,6 +178,13 @@ static ngx_command_t  ngx_http_ssl_comma
       offsetof(ngx_http_ssl_srv_conf_t, trusted_certificate),
       NULL },
 
+    { ngx_string("ssl_trusted_ca_path"),
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_CONF_TAKE1,
+      ngx_conf_set_str_slot,
+      NGX_HTTP_SRV_CONF_OFFSET,
+      offsetof(ngx_http_ssl_srv_conf_t, trusted_ca_path),
+      NULL },
+
     { ngx_string("ssl_prefer_server_ciphers"),
       NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_CONF_FLAG,
       ngx_conf_set_flag_slot,
@@ -603,6 +610,7 @@ ngx_http_ssl_create_srv_conf(ngx_conf_t
      *     sscf->ecdh_curve = { 0, NULL };
      *     sscf->client_certificate = { 0, NULL };
      *     sscf->trusted_certificate = { 0, NULL };
+     *     sscf->trusted_ca_path = { 0, NULL };
      *     sscf->crl = { 0, NULL };
      *     sscf->ciphers = { 0, NULL };
      *     sscf->shm_zone = NULL;
@@ -674,6 +682,7 @@ ngx_http_ssl_merge_srv_conf(ngx_conf_t *
                          "");
     ngx_conf_merge_str_value(conf->trusted_certificate,
                          prev->trusted_certificate, "");
+    ngx_conf_merge_str_value(conf->trusted_ca_path, prev->trusted_ca_path, "");
     ngx_conf_merge_str_value(conf->crl, prev->crl, "");
 
     ngx_conf_merge_str_value(conf->ecdh_curve, prev->ecdh_curve,
@@ -804,6 +813,7 @@ ngx_http_ssl_merge_srv_conf(ngx_conf_t *
 
     if (ngx_ssl_trusted_certificate(cf, &conf->ssl,
                                     &conf->trusted_certificate,
+                                    &conf->trusted_ca_path,
                                     conf->verify_depth)
         != NGX_OK)
     {
--- nginx.orig/src/http/modules/ngx_http_ssl_module.h
+++ nginx/src/http/modules/ngx_http_ssl_module.h
@@ -42,6 +42,7 @@ typedef struct {
     ngx_str_t                       ecdh_curve;
     ngx_str_t                       client_certificate;
     ngx_str_t                       trusted_certificate;
+    ngx_str_t                       trusted_ca_path;
     ngx_str_t                       crl;
 
     ngx_str_t                       ciphers;
--- nginx.orig/src/http/modules/ngx_http_uwsgi_module.c
+++ nginx/src/http/modules/ngx_http_uwsgi_module.c
@@ -53,6 +53,7 @@ typedef struct {
     ngx_str_t                  ssl_ciphers;
     ngx_uint_t                 ssl_verify_depth;
     ngx_str_t                  ssl_trusted_certificate;
+    ngx_str_t                  ssl_trusted_ca_path;
     ngx_str_t                  ssl_crl;
     ngx_array_t               *ssl_conf_commands;
 #endif
@@ -538,6 +539,13 @@ static ngx_command_t ngx_http_uwsgi_comm
       offsetof(ngx_http_uwsgi_loc_conf_t, ssl_trusted_certificate),
       NULL },
 
+    { ngx_string("uwsgi_ssl_trusted_ca_path"),
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_TAKE1,
+      ngx_conf_set_str_slot,
+      NGX_HTTP_LOC_CONF_OFFSET,
+      offsetof(ngx_http_uwsgi_loc_conf_t, ssl_trusted_ca_path),
+      NULL },
+
     { ngx_string("uwsgi_ssl_crl"),
       NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_TAKE1,
       ngx_conf_set_str_slot,
@@ -1895,6 +1903,8 @@ ngx_http_uwsgi_merge_loc_conf(ngx_conf_t
                               prev->ssl_verify_depth, 1);
     ngx_conf_merge_str_value(conf->ssl_trusted_certificate,
                               prev->ssl_trusted_certificate, "");
+    ngx_conf_merge_str_value(conf->ssl_trusted_ca_path,
+                              prev->ssl_trusted_ca_path, "");
     ngx_conf_merge_str_value(conf->ssl_crl, prev->ssl_crl, "");
 
     ngx_conf_merge_ptr_value(conf->upstream.ssl_certificate,
@@ -2576,14 +2586,15 @@ ngx_http_uwsgi_set_ssl(ngx_conf_t *cf, n
     }
 
     if (uwcf->upstream.ssl_verify) {
-        if (uwcf->ssl_trusted_certificate.len == 0) {
+        if (uwcf->ssl_trusted_certificate.len == 0 || uwcf->ssl_trusted_ca_path.len == 0) {
             ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
-                      "no uwsgi_ssl_trusted_certificate for uwsgi_ssl_verify");
+                      "no uwsgi_ssl_trusted_certificate or uwsgi_ssl_trusted_ca_path for uwsgi_ssl_verify");
             return NGX_ERROR;
         }
 
         if (ngx_ssl_trusted_certificate(cf, uwcf->upstream.ssl,
                                         &uwcf->ssl_trusted_certificate,
+                                        &uwcf->ssl_trusted_ca_path,
                                         uwcf->ssl_verify_depth)
             != NGX_OK)
         {
--- nginx.orig/src/mail/ngx_mail_ssl_module.c
+++ nginx/src/mail/ngx_mail_ssl_module.c
@@ -188,6 +188,14 @@ static ngx_command_t  ngx_mail_ssl_comma
       offsetof(ngx_mail_ssl_conf_t, trusted_certificate),
       NULL },
 
+
+    { ngx_string("ssl_trusted_ca_path"),
+      NGX_MAIL_MAIN_CONF|NGX_MAIL_SRV_CONF|NGX_CONF_TAKE1,
+      ngx_conf_set_str_slot,
+      NGX_MAIL_SRV_CONF_OFFSET,
+      offsetof(ngx_mail_ssl_conf_t, trusted_ca_path),
+      NULL },
+
     { ngx_string("ssl_crl"),
       NGX_MAIL_MAIN_CONF|NGX_MAIL_SRV_CONF|NGX_CONF_TAKE1,
       ngx_conf_set_str_slot,
@@ -303,6 +311,7 @@ ngx_mail_ssl_create_conf(ngx_conf_t *cf)
      *     scf->ecdh_curve = { 0, NULL };
      *     scf->client_certificate = { 0, NULL };
      *     scf->trusted_certificate = { 0, NULL };
+     *     scf->trusted_ca_path = { 0, NULL };
      *     scf->crl = { 0, NULL };
      *     scf->ciphers = { 0, NULL };
      *     scf->shm_zone = NULL;
@@ -366,6 +375,7 @@ ngx_mail_ssl_merge_conf(ngx_conf_t *cf,
                          prev->client_certificate, "");
     ngx_conf_merge_str_value(conf->trusted_certificate,
                          prev->trusted_certificate, "");
+    ngx_conf_merge_str_value(conf->trusted_ca_path, prev->trusted_ca_path, "");
     ngx_conf_merge_str_value(conf->crl, prev->crl, "");
 
     ngx_conf_merge_str_value(conf->ciphers, prev->ciphers, NGX_DEFAULT_CIPHERS);
@@ -466,6 +476,7 @@ ngx_mail_ssl_merge_conf(ngx_conf_t *cf,
 
         if (ngx_ssl_trusted_certificate(cf, &conf->ssl,
                                         &conf->trusted_certificate,
+                                        &conf->trusted_ca_path,
                                         conf->verify_depth)
             != NGX_OK)
         {
--- nginx.orig/src/mail/ngx_mail_ssl_module.h
+++ nginx/src/mail/ngx_mail_ssl_module.h
@@ -42,6 +42,7 @@ typedef struct {
     ngx_str_t        ecdh_curve;
     ngx_str_t        client_certificate;
     ngx_str_t        trusted_certificate;
+    ngx_str_t        trusted_ca_path;
     ngx_str_t        crl;
 
     ngx_str_t        ciphers;
--- nginx.orig/src/stream/ngx_stream_proxy_module.c
+++ nginx/src/stream/ngx_stream_proxy_module.c
@@ -46,6 +46,7 @@ typedef struct {
     ngx_flag_t                       ssl_verify;
     ngx_uint_t                       ssl_verify_depth;
     ngx_str_t                        ssl_trusted_certificate;
+    ngx_str_t                        ssl_trusted_ca_path;
     ngx_str_t                        ssl_crl;
     ngx_stream_complex_value_t      *ssl_certificate;
     ngx_stream_complex_value_t      *ssl_certificate_key;
@@ -320,6 +321,13 @@ static ngx_command_t  ngx_stream_proxy_c
       offsetof(ngx_stream_proxy_srv_conf_t, ssl_trusted_certificate),
       NULL },
 
+    { ngx_string("proxy_ssl_trusted_ca_path"),
+      NGX_STREAM_MAIN_CONF|NGX_STREAM_SRV_CONF|NGX_CONF_TAKE1,
+      ngx_conf_set_str_slot,
+      NGX_STREAM_SRV_CONF_OFFSET,
+      offsetof(ngx_stream_proxy_srv_conf_t, ssl_trusted_ca_path),
+      NULL },
+
     { ngx_string("proxy_ssl_crl"),
       NGX_STREAM_MAIN_CONF|NGX_STREAM_SRV_CONF|NGX_CONF_TAKE1,
       ngx_conf_set_str_slot,
@@ -2070,6 +2078,7 @@ ngx_stream_proxy_create_srv_conf(ngx_con
      *     conf->ssl_protocols = 0;
      *     conf->ssl_ciphers = { 0, NULL };
      *     conf->ssl_trusted_certificate = { 0, NULL };
+     *     conf->ssl_trusted_ca_path = { 0, NULL };
      *     conf->ssl_crl = { 0, NULL };
      *
      *     conf->ssl = NULL;
@@ -2181,6 +2190,9 @@ ngx_stream_proxy_merge_srv_conf(ngx_conf
     ngx_conf_merge_str_value(conf->ssl_trusted_certificate,
                               prev->ssl_trusted_certificate, "");
 
+    ngx_conf_merge_str_value(conf->ssl_trusted_ca_path,
+                              prev->ssl_trusted_ca_path, "");
+
     ngx_conf_merge_str_value(conf->ssl_crl, prev->ssl_crl, "");
 
     ngx_conf_merge_ptr_value(conf->ssl_certificate,
@@ -2314,14 +2326,15 @@ ngx_stream_proxy_set_ssl(ngx_conf_t *cf,
     }
 
     if (pscf->ssl_verify) {
-        if (pscf->ssl_trusted_certificate.len == 0) {
+        if (pscf->ssl_trusted_certificate.len == 0 || pscf->ssl_trusted_ca_path.len == 0) {
             ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
-                      "no proxy_ssl_trusted_certificate for proxy_ssl_verify");
+                      "no proxy_ssl_trusted_certificate or proxy_ssl_trusted_ca_path for proxy_ssl_verify");
             return NGX_ERROR;
         }
 
         if (ngx_ssl_trusted_certificate(cf, pscf->ssl,
                                         &pscf->ssl_trusted_certificate,
+                                        &pscf->ssl_trusted_ca_path,
                                         pscf->ssl_verify_depth)
             != NGX_OK)
         {
--- nginx.orig/src/stream/ngx_stream_ssl_module.c
+++ nginx/src/stream/ngx_stream_ssl_module.c
@@ -170,6 +170,13 @@ static ngx_command_t  ngx_stream_ssl_com
       offsetof(ngx_stream_ssl_srv_conf_t, trusted_certificate),
       NULL },
 
+    { ngx_string("ssl_trusted_ca_path"),
+      NGX_STREAM_MAIN_CONF|NGX_STREAM_SRV_CONF|NGX_CONF_TAKE1,
+      ngx_conf_set_str_slot,
+      NGX_STREAM_SRV_CONF_OFFSET,
+      offsetof(ngx_stream_ssl_srv_conf_t, trusted_ca_path),
+      NULL },
+
     { ngx_string("ssl_prefer_server_ciphers"),
       NGX_STREAM_MAIN_CONF|NGX_STREAM_SRV_CONF|NGX_CONF_FLAG,
       ngx_conf_set_flag_slot,
@@ -859,6 +866,7 @@ ngx_stream_ssl_merge_srv_conf(ngx_conf_t
                          "");
     ngx_conf_merge_str_value(conf->trusted_certificate,
                          prev->trusted_certificate, "");
+    ngx_conf_merge_str_value(conf->trusted_ca_path, prev->trusted_ca_path, "");
     ngx_conf_merge_str_value(conf->crl, prev->crl, "");
     ngx_conf_merge_str_value(conf->alpn, prev->alpn, "");
 
@@ -971,6 +979,7 @@ ngx_stream_ssl_merge_srv_conf(ngx_conf_t
 
         if (ngx_ssl_trusted_certificate(cf, &conf->ssl,
                                         &conf->trusted_certificate,
+                                        &conf->trusted_ca_path,
                                         conf->verify_depth)
             != NGX_OK)
         {
--- nginx.orig/src/stream/ngx_stream_ssl_module.h
+++ nginx/src/stream/ngx_stream_ssl_module.h
@@ -41,6 +41,7 @@ typedef struct {
     ngx_str_t        ecdh_curve;
     ngx_str_t        client_certificate;
     ngx_str_t        trusted_certificate;
+    ngx_str_t        trusted_ca_path;
     ngx_str_t        crl;
     ngx_str_t        alpn;
 
